### DISCUSSION TITLE
Batch recursive (nested) updates

### DISCUSSION THREAD

**Discussion by crabvk** [ai/nanostores#264] [Q&A]
I have a tree data model where each node has a boolean state, i.e. `expanded`. So, I'd define it like this:
```typescript
interface TreeNode {
  expanded: Atom<boolean>
  children: TreeNode[]
}
```
(Don't know the actual type of value `atom` function returns, so just using a generic pseudo `Atom` type).
Now I want to define `expandAll` function that would take a root node and set all the nested node states to `true`:
```typescript
function expandAll(parentNode: TreeNode) {
  for (const node of parentNode.children) {
    expandAll(node)
  }
  parentNode.expanded.set(true)
}
```
How to batch such updates? I don't want my TreeView component to re-render until all nodes expanded.

**Comment by euaaaio** [comment]
It depends on rendering implementation and optimizations of your framework.
For instance, if you will use Vue `useStore`, Nano Stores will trigger updates one by one, but Vue will merge them into one rerender.

So the best practice is to set new value for entire store once, instead of changing children items one by one.

**Comment by ai** [comment]
The best way is to batch on UI framework side. React, Vue, Svelte do it.

**Comment by crabvk** [comment]
I'm using React.
The dev (collaborator) of preactjs/signals [says](https://github.com/preactjs/signals/discussions/501) React batching is not reliable (has no guarantee on how and when React will batch state updates)

**Reply by ai** [reply]
I had no issues with React batching.

In Preact talk discussion I don't see a wrong output in the example, only an unnecessary render which is not a big problem.
